### PR TITLE
[VQueues] Fix iterator being blind in edge cases

### PR DESCRIPTION
**Pull request by AhmedSoliman**
- Adds a test to validate the corner case.
- Invalidate the iterator when we observe items > head (temporary fix).

---
[//]: # (BEGIN SAPLING FOOTER)
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/restatedev/restate/pull/4652).
* #4635
* #4644
* #4649
* #4648
* #4647
* __->__ #4652
* #4646
* #4643
* #4642

### CODE REVIEW

File: crates/partition-store/src/tests/vqueue_table_test/mod.rs [hunk 1]
```diff
@@ -632,6 +632,190 @@ async fn tailing_iterator_sees_inserted_ahead_without_reseek(rocksdb: &mut Parti
...
+    let entry6 = entry(6, false, 60, 6);
+    let entry9 = entry(9, false, 90, 9);
+    {
+        let mut txn = rocksdb.transaction();
+        txn.put_vqueue_inbox(&qid, Stage::Inbox, &entry1.0, &entry1.1);
+        txn.put_vqueue_inbox(&qid, Stage::Inbox, &entry3.0, &entry3.1);
+        txn.put_vqueue_inbox(&qid, Stage::Inbox, &entry6.0, &entry6.1);
+        txn.put_vqueue_inbox(&qid, Stage::Inbox, &entry9.0, &entry9.1);
+        txn.commit().await.expect("commit should succeed");
+    }
```
Comment (tillrohrmann) [inline crates/partition-store/src/tests/vqueue_table_test/mod.rs:653]: I think the or a flush should be here so that these entries are in a SST file.

File: crates/partition-store/src/tests/vqueue_table_test/mod.rs [hunk 2]
```diff
@@ -632,6 +632,190 @@ async fn tailing_iterator_sees_inserted_ahead_without_reseek(rocksdb: &mut Parti
...
+
+    let db = rocksdb.partition_db();
+    let mut reader = db.new_inbox_reader(&qid);
+
+    // Single seek before inserting new items: land on entry3.
+    reader.seek_after(&qid, &entry1.0);
```
Comment (tillrohrmann) [inline crates/partition-store/src/tests/vqueue_table_test/mod.rs]: Whether the flush happens after seek of after  line 715 is a good question. But it should happen before committing `entry4`.

File: crates/partition-store/src/tests/vqueue_table_test/mod.rs [hunk 3]
```diff
@@ -632,6 +632,190 @@ async fn tailing_iterator_sees_inserted_ahead_without_reseek(rocksdb: &mut Parti
...
+        txn.put_vqueue_inbox(&qid, Stage::Inbox, &entry6.0, &entry6.1);
+        txn.put_vqueue_inbox(&qid, Stage::Inbox, &entry9.0, &entry9.1);
+        txn.commit().await.expect("commit should succeed");
+    }
+
+    let db = rocksdb.partition_db();
+    let mut reader = db.new_inbox_reader(&qid);
+
+    // Single seek before appending new tail items: land on entry3.
+    reader.seek_after(&qid, &entry1.0);
```
Comment (tillrohrmann) [inline crates/partition-store/src/tests/vqueue_table_test/mod.rs:799]: Same here.

File: crates/partition-store/src/tests/vqueue_table_test/mod.rs [hunk 4]
```diff
@@ -630,6 +630,205 @@ async fn tailing_iterator_sees_inserted_ahead_without_reseek(rocksdb: &mut Parti
...
+        remaining_ids,
+        vec![entry_id(4), entry_id(5), entry_id(6), entry_id(9)],
+        "Spliced flushed items should be visible and not skipped without re-seek"
+    );
+}
+
+/// Test: Seeked tailing iterator sees spliced insertions after a pre-insert flush.
+///
+/// Scenario: perform one seek to position the cursor on the item immediately
+/// before the splice point, flush memtables, then insert a new item in the middle.
```
Comment (tillrohrmann) [inline crates/partition-store/src/tests/vqueue_table_test/mod.rs:697]: Description is off. We first flush and then seek.

File: crates/partition-store/src/tests/vqueue_table_test/mod.rs [hunk 5]
```diff
@@ -630,6 +630,205 @@ async fn tailing_iterator_sees_inserted_ahead_without_reseek(rocksdb: &mut Parti
...
+        let mut txn = rocksdb.transaction();
+        txn.put_vqueue_inbox(&qid, Stage::Inbox, &entry4.0, &entry4.1);
+        txn.commit().await.expect("commit should succeed");
+    }
+
+    // Continue from the seeked position without another seek.
+    reader.advance();
+
+    // This is the proof that we need to re-seek because the iterator will be blind
+    // to the newly added item if the seek happened after the mutable memtable flush.
```
Comment (tillrohrmann) [inline crates/partition-store/src/tests/vqueue_table_test/mod.rs:742]: I think a seek after the memtable flush is not the only sufficient condition. The problem also occurs w/o the seek. The following test fails:

```
async fn tailing_iterator_sees_flushed_insertions_mid_iteration(rocksdb: &mut PartitionStore) {
    let qid = VQueueId::custom(7_250, "1");

    let entry1 = entry(1, false, 10, 1);
    let entry3 = entry(3, false, 30, 3);
    let entry6 = entry(6, false, 60, 6);
    let entry9 = entry(9, false, 90, 9);
    {
        let mut txn = rocksdb.transaction();
        txn.put_vqueue_inbox(&qid, Stage::Inbox, &entry1.0, &entry1.1);
        txn.put_vqueue_inbox(&qid, Stage::Inbox, &entry3.0, &entry3.1);
        txn.put_vqueue_inbox(&qid, Stage::Inbox, &entry6.0, &entry6.1);
        txn.put_vqueue_inbox(&qid, Stage::Inbox, &entry9.0, &entry9.1);
        txn.commit().await.expect("commit should succeed");
    }

    let db = rocksdb.partition_db();
    let mut reader = db.new_inbox_reader(&qid);
    reader.seek_to_first();

    assert_eq!(*reader.peek().unwrap().unwrap().0.entry_id(), entry_id(1));

    rocksdb
        .partition_db()
        .flush_memtables(true)
        .await
        .expect("flush memtables should succeed");

    reader.advance();
    assert_eq!(*reader.peek().unwrap().unwrap().0.entry_id(), entry_id(3));

    // Splice entries between current position (3) and next existing item (6).
    let entry4 = entry(4, false, 40, 4);
    let entry5 = entry(5, false, 50, 5);
    {
        let mut txn = rocksdb.transaction();
        txn.put_vqueue_inbox(&qid, Stage::Inbox, &entry4.0, &entry4.1);
        txn.put_vqueue_inbox(&qid, Stage::Inbox, &entry5.0, &entry5.1);
        txn.commit().await.expect("commit should succeed");
    }

    // Continue from current position without seek.
    reader.advance();
    let mut remaining_ids = Vec::new();
    while let Ok(Some(item)) = reader.peek() {
        remaining_ids.push(*item.0.entry_id());
        reader.advance();
    }

    // fails seeing entry_id(4) and entry_id(5)
    assert_eq!(
        remaining_ids,
        vec![entry_id(4), entry_id(5), entry_id(6), entry_id(9)],
        "Spliced flushed items should be visible and not skipped without re-seek"
    );
}
```

File: crates/vqueues/src/scheduler/queue.rs [hunk 6]
```diff
@@ -135,6 +134,11 @@ impl<S: VQueueStore> Queue<S> {
                     // Ensure that next advance would re-seek to the newly added item
                     self.reader = Reader::Closed;
                     return true;
+                } else {
+                    // This is a temporary fix to ensure that we perform a re-seek
+                    // to fix the issue where the iterator wouldn't see the newly added
+                    // items if the memtable was flushed prior the seek.
```
Comment (tillrohrmann) [inline crates/vqueues/src/scheduler/queue.rs:140]: Seeking after a flush is one sufficient condition but not the only one. Advancing the iterator is sufficient as well in some cases.

**Review by tillrohrmann — COMMENTED** [review]
Thanks a lot for fixing this issue @AhmedSoliman. Now the e2e-verification-runner tests should hopefully be green again :-) 

I think the problematic behavior is not strictly tied to seeking after a flush as it can also happen with advancing the iterator after a flush.

**Review by tillrohrmann — COMMENTED** [review]
I am wondering whether the condition in `VqueueState::is_dormant()` (https://github.com/restatedev/restate/blob/1d0d5c48f8011a40af7ce53a49941568a106fd16/crates/vqueues/src/scheduler/vqueue_state.rs#L334) is correct. We are marking a queue as dormant if `queue.is_empty()` and there are no unconfirmed assignments. This might cause the drr to remove the queue from its internal cache. If now `!meta.is_inbox_empty()`, then `VQueue::enqueue_new` won't add the vqueue back via `EventDetails::AddQueue` and we might panic as seen in the e2e verification test. Hence, I was thinking whether the `queue.is_empty()` check should rather be an assertion like: `meta.is_inbox_empty()` iff `queue.is_empty()`.
